### DISCUSSION TITLE
STL does not retain manifoldness

### DISCUSSION THREAD

**Discussion by briansturgill** [elalish/manifold#243] [Q&A]
This isn't a huge deal to me, but I thought you would want to know about it.
Meshlab reports the output stl file as having 1 non-manifold vertex.
It is of course, the point where the two spheres touch.
It is an edge case, but technically it is non-manifold.
In practice I doubt it hurts anything, though I assume the resulting Manifold wouldn't be allowed in further boolean ops?
```python
from pymanifold import Manifold
import trimesh

def save(name, model):
	mesh = model.to_mesh()
	meshOut = trimesh.Trimesh(
		vertices=mesh.vert_pos, faces=mesh.tri_verts)
	trimesh.exchange.export.export_mesh(meshOut, f'{name}.stl', 'stl')
	print(f'Exported model to {name}.stl')
    
segments = 150

s1 = Manifold.sphere(radius=1, circular_segments=150)
s2 = Manifold.sphere(radius=1, circular_segments=150).translate(2, 0, 0)
ret = s1+s2
save("union-touching", ret)
```

**Comment by elalish** [comment]
The trouble in this case is the STL format. Since it is triangle soup without connectivity information, if forces all downstream apps to merge vertices and hope, which fails in this case and many others. If you care about manifoldness, you'll need to use basically any format but STL (I'd recommend 3MF, since I helped spec it, but glTF, even OBJ work just fine).

**Comment by briansturgill** [comment]
@elalish 
Well, I tried obj, xyz, ply, off, and glb and all are showing 1 point non-manifold in meshlab.
Few things support 3mf.

**Reply by elalish** [reply]
Oh, I read your comment wrong, sorry. The definition of manifold I'm using is edge-only. I've been wondering if there's a need to remove touching verts. Shouldn't be too hard. Know of any reasons they are problematic?

**Reply by briansturgill** [reply]
So long as one is able to use the generated touching mesh inside Manifold, I don't know of a problem.
Certainly PrusaSlicer accepts it with no problems.

**Reply by elalish** [reply]
I actually just tried your example, but investigating the manifold internally instead of in Meshlab. It turns out our library doesn't even produce the non-manifold vertex. It correctly produces a duplicate vertex - same position, but one is in one sphere and the other is in the other. I would bet whatever library is being used either to export the model to a file, or what Meshlab does when reading that file is collapsing that duplicate vertex and creating the non-manifold point. This is very common because of how bad and ubiquitous the STL format was for so long, but if we actually care about manifoldness, one cannot use geometric nearness to assume topological connection.

**Comment by pentacular** [comment]
I think this means that it allows zero thickness parts, meaning it's a form of self-intersection.

I think also that implies that it allows negative thickness parts, due to how rounding errors can accumulate.

So this discussion should probably be about self-intersection rather than manifoldness.

**Reply by elalish** [reply]
Indeed, zero-thickness parts are allowed, though our library does a pretty good job of not generating them in the first place (good symbolic perturbation) and removing them where possible (through decimation). And due to accumulation of rounding errors we use a very specific definition of self-intersection that guides our algorithms: [&epsilon;-valid](https://github.com/elalish/manifold/wiki/Manifold-Library#overlaps). And we track this &epsilon; through the operations with `Precision()`.
